internal/grpcsync/pubsub_test.go

```diff
@@ -0,0 +1,119 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpcsync
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+type mockSubscriber struct {
+	mu   sync.Mutex
+	msgs []string
+	wg   *sync.WaitGroup
+}
+
+func (ms *mockSubscriber) OnMessage(msg interface{}) {
+	ms.mu.Lock()
+	defer ms.mu.Unlock()
+	ms.msgs = append(ms.msgs, msg.(string))
+	ms.wg.Done()
+}
+
+func (s) TestPubSub(t *testing.T) {
```

Comment:
I feel that it would be nice to have separate tests for each of the scenarios that we want to test:
- Tests around `Subscribe` (each of these tests could be run with a single subscriber *and* with multiple subscribers using `t.Run()`)
  - Create a pubsub with no messages. Register a subscriber. Ensure that the callback is not invoked within a reasonable amount of time. We generally use `10ms` for things like this where we wait for an event that we expect *not* to happen. Please take a look at tests in `internal/grpcsync/callback_serializer_test.go` to get an idea of how we generally do this.
  - Create a pubsub. Pushlish one message. Register a subscriber. Ensure that the callback is invoked with the published mesage.
  - Create a pubsub. Pushing more than one message. Register a subscriber. Ensure that the callback is invoked only once and it is invoked with the most recent message.
  - Create a pubsub with no messages. Register a subscriber. Publish multiple messages and ensure that the subscriber receives them in order. 
  - Create a pubsub with no messages. Register a subscriber. Publish multiple messages. Ensure that the subscriber received them in order. Stop the pubsub. Push one more message. Ensure that the subscribers dont receive this last message.
  
  These are just the scenarios that came to my mind currently. I'm pretty sure, there could be more interesting ones. I'll leave that to you.

Reply:
Having more than one subscriber, and unregistering one of them, and ensuring that a new message is sent to all subscribers except the unregistered one is also a good case to test.